base/operators.jl

```diff
@@ -157,6 +157,27 @@ isless(x::AbstractFloat, y::AbstractFloat) = (!isnan(x) & (isnan(y) | signless(x
 isless(x::Real,          y::AbstractFloat) = (!isnan(x) & (isnan(y) | signless(x, y))) | (x < y)
 isless(x::AbstractFloat, y::Real         ) = (!isnan(x) & (isnan(y) | signless(x, y))) | (x < y)
 
+"""
+    isgreater(x, y)
+
+Test whether `x` is greater than `y`, according to a fixed total order.
+`isgreater` is defined in terms of `isless`, but is not the opposite of that function.
+
+`isless` defines a fixed total order that ascends, while `isgreater` defines a
+fixed total order that descends. Both order values that are normally unordered,
+such as `NaN`, after all regular values and [`missing`](@ref) last. So for
+`isless` these values are biggest and for `isgreater` these values are
+smallest.
+
+# Implementation
+Types should not usually implement this function. Instead, implement `isless`.
+"""
+isgreater(x, y) = _is_reflexive(x) && _is_reflexive(y) ? isless(y, x) : isless(x, y)
```

Comment:
Should that be `!isless(x, y)` in the final branch?

Reply:
I don't think so. That's the branch for NaN and missing and I just want those sorted the same as for isless.

Reply:
This definition is the trickiest part of this PR and I think it deserves _at least_ some comments about it in the source code.

Having said that, I still have a very uneasy feeling about this definition.  In particular, I _think_ proving that `isgreater` satisfies transitivity requires that "`_is_reflexive(x) && !_is_reflexive(y)` implies `isless(x, y)`" holds.  This is the case for `NaN` and `missing` but this is not in the API.  It'd be much easier to understand and enforce that `isgreater` is a total order [*] if we just use

```julia
function isgreater2(x, y)
    xisnan = x != x
    xisnan isa Bool || return false  # x is missing
    yisnan = y != y
    yisnan isa Bool || return true  # y is missing
    xisnan && return false
    yisnan && return true
    return isless(y, x)
end
```

In other words, using `_is_reflexive` may look like it makes `isgreater` very generic but it crucially depends on the implementation details.

Anyway, I think this PR has to do at least one of the following:

* Use `isgreater2` (or something equivalent) instead of current `isgreater`.
* Add the aforementioned property of `=`-reflexivity and `isless` to the part of implementation requirement of `=` and `isless`.

(Note: Transitivity is very important as otherwise "min" is not going to be associative. If "min" is not associative, we can't have parallel argmin/findmin/minimum/extrema.)

[*] OK this is not quite enough for user-defined `missing`-like types but we don't have such facility even for `isless` yet.